### PR TITLE
feat: use container DNS when in container mode

### LINKED ISSUE #8303: CoreDNS fails when using Talos in Docker with docker default nameserver
### Description

We were just debugging some DNS issues when running Talos locally as a docker container. Since `8.8.8.8` is not accessible inside our network, we implemented a fix to use the locally available DNS in `/etc/resolv`. Unfortunately, this does not work on Linux machines since there is a DNS resolver hosted on `127.0.0.1`. When [trying to use the docker dns `127.0.0.11`](https://medium.com/@prajwal.chin/understanding-docker-dns-2ed4b070a0#:~:text=By%20default%2C%20Docker%20uses%20the,0.11), we ended up with an infinite dns loop. It looks like [kind is aware of that and tries to mitigate it](https://hwchiu.medium.com/fun-dns-facts-learned-from-the-kind-environment-241e0ea8c6d4), could talos do the same?

Thank you!
### Logs
```
❯ kubectl -n kube-system logs coredns-565ffcd8f4-hnmq6
.:53
[INFO] plugin/reload: Running configuration SHA512 = f43368fe881b6cd37b121f37ba0b71c065df5bfc99b5c5c05d7f95bf82289d7ab7e78d5b98c1f02172d8004a8a8f34027cef04e86c780d40a7c5d1301559f5b3
CoreDNS-1.11.1
linux/amd64, go1.20.7, ae2bbc2
[FATAL] plugin/loop: Loop (127.0.0.1:35656 -> :53) detected for zone ".", see https://coredns.io/plugins/loop#troubleshooting. Query: "HINFO 1776841706389146686.271906425746229838."
```
### Environment

- Talos version: v1.6.4
- Platform: Docker

### PR DESCRIPTION
**Pull request by smira**
More specifically, pick up `/etc/resolv.conf` contents by default when in container mode, and use that as a base resolver for the host DNS.

Fixes #8303

### CODE REVIEW

File: internal/app/machined/pkg/runtime/v1alpha1/platform/container/container.go [hunk 1]
```diff
@@ -77,6 +78,34 @@ func (c *Container) NetworkConfiguration(ctx context.Context, _ state.State, ch
...
+	resolverSpec := network.ResolverSpecSpec{
+		ConfigLayer: network.ConfigPlatform,
+	}
+
+	resolvers, err := os.ReadFile("/etc/resolv.conf")
+	if err != nil {
+		return err
+	}
+
+	for _, line := range bytes.Split(resolvers, []byte("\n")) {
```
Comment (DmitriyMV) [inline internal/app/machined/pkg/runtime/v1alpha1/platform/container/container.go:90]: Maybe use `bufio.NewScanner(f)` and `f, err := os.Open("/etc/resolv.conf")`? 

That way there is no need for `[]byte(...)` conversions.

Comment (smira) [reply internal/app/machined/pkg/runtime/v1alpha1/platform/container/container.go:90]: let's call it PoC for now, I think this way is actually easier for a small file.

File: internal/app/machined/pkg/controllers/cluster/local_affiliate_test.go [hunk 2]
```diff
@@ -169,7 +169,10 @@ func (suite *LocalAffiliateSuite) TestCPGeneration() {
 		asrt.Equal("Talos ("+version.Tag+")", spec.OperatingSystem)
 		asrt.Equal(cluster.KubeSpanAffiliateSpec{}, spec.KubeSpan)
 		asrt.NotNil(spec.ControlPlane)
-		asrt.Equal(6445, spec.ControlPlane.APIServerPort)
+
+		if spec.ControlPlane != nil {
+			asrt.Equal(6445, spec.ControlPlane.APIServerPort)
+		}
```
Comment (DmitriyMV) [inline internal/app/machined/pkg/controllers/cluster/local_affiliate_test.go:175]: IMHO this looks weird given the check above. I think we can actually do:

```suggestion
		asrt.Equal(6445, pointer.SafeDeref(spec.ControlPlane).APIServerPort)
```

File: pkg/machinery/config/generate/worker.go [hunk 3]
```diff
@@ -96,10 +96,14 @@ func (in *Input) worker() ([]config.Document, error) {
...
 
-	if in.Options.VersionContract.LocalDNSEnabled() {
+	if in.Options.VersionContract.HostDNSEnabled() || in.Options.HostDNSForwardKubeDNSToHost.ValueOrZero() {
 		machine.MachineFeatures.HostDNSSupport = &v1alpha1.HostDNSConfig{
 			HostDNSEnabled: pointer.To(true),
 		}
+
+		if in.Options.HostDNSForwardKubeDNSToHost.ValueOrZero() {
+			machine.MachineFeatures.HostDNSSupport.HostDNSForwardKubeDNSToHost = pointer.To(true)
+		}
```
Comment (DmitriyMV) [inline pkg/machinery/config/generate/worker.go:106]: I think we can set the value directly in `HostDNSSupport`:

```go
		machine.MachineFeatures.HostDNSSupport = &v1alpha1.HostDNSConfig{
			HostDNSEnabled: pointer.To(true),
			HostDNSForwardKubeDNSToHost: in.Options.HostDNSForwardKubeDNSToHost.Ptr(),
		}
```

File: pkg/machinery/config/generate/init.go [hunk 4]
```diff
@@ -95,10 +95,14 @@ func (in *Input) init() ([]config.Document, error) {
...
 
-	if in.Options.VersionContract.LocalDNSEnabled() {
+	if in.Options.VersionContract.HostDNSEnabled() || in.Options.HostDNSForwardKubeDNSToHost.ValueOrZero() {
 		machine.MachineFeatures.HostDNSSupport = &v1alpha1.HostDNSConfig{
 			HostDNSEnabled: pointer.To(true),
 		}
+
+		if in.Options.HostDNSForwardKubeDNSToHost.ValueOrZero() {
+			machine.MachineFeatures.HostDNSSupport.HostDNSForwardKubeDNSToHost = pointer.To(true)
+		}
```
Comment (DmitriyMV) [inline pkg/machinery/config/generate/init.go:105]: I think we can set the value directly in `HostDNSSupport`:

```go
		machine.MachineFeatures.HostDNSSupport = &v1alpha1.HostDNSConfig{
			HostDNSEnabled: pointer.To(true),
			HostDNSForwardKubeDNSToHost: in.Options.HostDNSForwardKubeDNSToHost.Ptr(),
		}
```

**Review by DmitriyMV — COMMENTED** [review]
I'm trying to understand and failing. 

So if we run Talos in Docker container we use container passed `resolv.conf`. **But** we require user to set `forwardKubeDNSToHost: true` which means we pass our own DNS to pods and use it for networking. How does this works together? Sorry if missing something obvious.

File: pkg/machinery/config/generate/worker.go (hunk 3 above)
Comment (smira) [reply pkg/machinery/config/generate/worker.go:106]: yep, great idea!

**Comment by smira** [comment]
> So if we run Talos in Docker container we use container passed `resolv.conf`. **But** we require user to set `forwardKubeDNSToHost: true` which means we pass our own DNS to pods and use it for networking. How does this works together? Sorry if missing something obvious.

so the idea is the following:

1. `/etc/resolv.conf` on container boot contains container runtime provided resolvers, which we want to read and use 
2. we could rewrite `/etc/resolv.conf` with HostDNS, but that would require synchronization between reading and re-writing, and might be more confusing for someone doing `read /etc/resolv.conf`, so on the host we use just container-runtime provided resolvers; we could rewrite `/etc/resolv.conf` though, it would work same way
3. we enforce `forwardKubeDNSToHost` so that CoreDNS in Kubernetes uses host DNS to passthrough requests to container runtime resolver. if we don't do it, e.g. Docker default '127.0.0.11' won't work from CoreDNS, but '10.96.0.9' (host DNS) would perfectly work

Hope I explained it.

File: internal/app/machined/pkg/controllers/config/acquire.go [hunk 5]
```diff
@@ -214,6 +214,12 @@ func (validationModeDiskConfig) RequiresInstall() bool {
 	return false
 }
 
+// RequiresInstall implements validation.RuntimeMode interface.
+func (validationModeDiskConfig) InContainer() bool {
```
Comment (Unix4ever) [inline internal/app/machined/pkg/controllers/config/acquire.go:218]:
```suggestion
// InContainer implements validation.RuntimeMode interface.
func (validationModeDiskConfig) InContainer() bool {
```

**Review by Unix4ever — APPROVED** [review]

File: internal/app/machined/pkg/runtime/v1alpha1/platform/container/internal/files/resolv.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		ConfigLayer: network.ConfigPlatform,
+	}
+
+	resolvers, err := os.ReadFile(path)
+	if err != nil {
+		return resolverSpec, err
+	}
+
+	for _, line := range bytes.Split(resolvers, []byte("\n")) {
+		log.Printf("line orig = %s", string(line))
```
Comment (frezbo) [inline internal/app/machined/pkg/runtime/v1alpha1/platform/container/internal/files/resolv.go:28]: needed?

**Review by frezbo — APPROVED** [review]

File: internal/app/machined/pkg/runtime/v1alpha1/platform/container/internal/files/resolv.go (hunk 6 above)
Comment (smira) [reply internal/app/machined/pkg/runtime/v1alpha1/platform/container/internal/files/resolv.go]: thanks, dropped!

File: internal/app/machined/pkg/controllers/config/acquire.go (hunk 5 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/config/acquire.go]: thanks, fixed!

**Review by DmitriyMV — APPROVED** [review]

**Pull request merged**